controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,177 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONTROLLER_INTERFACE__ASYNC_FUNCTION_HANDLER_HPP_
+#define CONTROLLER_INTERFACE__ASYNC_FUNCTION_HANDLER_HPP_
+
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "lifecycle_msgs/msg/state.hpp"
+#include "rclcpp/duration.hpp"
+#include "rclcpp/time.hpp"
+#include "rclcpp_lifecycle/state.hpp"
+
+namespace ros2_control
+{
+template <typename T>
+class AsyncFunctionHandler
+{
+public:
+  AsyncFunctionHandler() = default;
+
+  ~AsyncFunctionHandler() { preempt_async_update(); }
+
+  void init(
+    std::function<const rclcpp_lifecycle::State &()> get_state_function,
+    std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function)
+  {
+    if (get_state_function == nullptr || async_function == nullptr)
+    {
+      throw std::runtime_error("AsyncFunctionHandler parsed functions are not valid!");
+    }
+    if (thread_.joinable())
+    {
+      throw std::runtime_error("AsyncFunctionHandler already initialized!");
+    }
+    get_state_function_ = get_state_function;
+    async_function_ = async_function;
+  }
+
+  T trigger_async_update(const rclcpp::Time & time, const rclcpp::Duration & period)
+  {
+    initialize_async_update_thread(get_state_function_, async_function_);
+    std::unique_lock<std::mutex> lock(async_mtx_, std::try_to_lock);
+    if (!async_update_ready_ && lock.owns_lock())
+    {
+      async_update_ready_ = true;
+      current_update_time_ = time;
+      current_update_period_ = period;
+      lock.unlock();
+      async_update_condition_.notify_one();
+    }
+    return async_update_return_;
+  }
+
+  void wait_for_update_to_finish()
+  {
+    if (is_running())
+    {
+      std::unique_lock<std::mutex> lock(async_mtx_);
+      async_update_condition_.wait(lock, [this] { return !async_update_ready_; });
+    }
+  }
+
+  bool is_initialized() const
+  {
+    return get_state_function_ != nullptr && async_function_ != nullptr;
+  }
+
+  void join_async_update_thread()
+  {
+    if (is_running())
+    {
+      thread_.join();
+    }
+  }
+
+  bool is_running() const { return thread_.joinable(); }
+
+  bool is_preempted() const
+  {
+    return (
+      async_update_stop_ ||
+      (is_initialized() &&
+       get_state_function_().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE));
+  }
+
+  void preempt_async_update()
```

Comment:
I'd call this `stop_async_update()`, since in the context of threading preempt means something different. Preemption is not a final act, it can occur frequently.

Reply:
Done